collect/central_collector.go

```diff
@@ -119,18 +153,235 @@ func (c *CentralCollector) collect() {
-	// add the span to the cache
-	trace := c.SpanCache.Get(sp.TraceID)
-	if trace == nil {
-		err := c.SpanCache.Set(sp)
-		if err != nil {
-			c.Logger.Error().Logf("error adding span with trace ID %s to cache: %s", sp.TraceID, err)
-			return
+func (c *CentralCollector) processTraces() error {
+	for {
+		select {
+		case <-c.done:
+			return nil
+		default:
+			ids := c.SpanCache.GetTraceIDs(processTracesBatchSize)
+			if len(ids) == 0 {
+				continue
+			}
+
+			statuses, err := c.Store.GetStatusForTraces(context.Background(), ids)
+			if err != nil {
+				c.Logger.Error().Logf("error getting statuses for traces: %s", err)
+				continue
+			}
+			for _, status := range statuses {
+				switch status.State {
+				case centralstore.DecisionKeep:
+					c.send(status)
+					c.SpanCache.Remove(status.TraceID)
+
+				case centralstore.DecisionDrop:
+					c.SpanCache.Remove(status.TraceID)
+				default:
+					c.Logger.Debug().Logf("trace %s is still pending", status.TraceID)
+				}
+			}
+		}
+
+		select {
+		case <-c.done:
+			return nil
+		default:
+		}
+
+		ticker := c.Clock.NewTicker(processTracesBackoffInterval)
+		select {
+		case <-c.done:
+			return nil
+		case <-ticker.Chan():
+			ticker.Stop()
+			continue
 		}
 	}
+}
 
-	trace = c.SpanCache.Get(sp.TraceID)
+func (c *CentralCollector) decide() error {
+	for {
+		select {
+		case <-c.done:
+			return nil
+		default:
+			if c.BlockOnDecider {
+				continue
+			}
+			ctx := context.Background()
+			tracesIDs, err := c.Store.GetTracesNeedingDecision(ctx, deciderBatchSize)
+			if err != nil {
+				c.Logger.Error().Logf("error getting traces needing decision: %s", err)
+				continue
+			}
+
+			if len(tracesIDs) == 0 {
+				continue
+			}
+			statuses, err := c.Store.GetStatusForTraces(ctx, tracesIDs)
+			if err != nil {
+				c.Logger.Error().Logf("error getting statuses for traces: %s", err)
+				continue
+			}
+			traces := make([]*centralstore.CentralTrace, len(statuses))
+			stateMap := make(map[string]*centralstore.CentralTraceStatus, len(statuses))
+			var wg sync.WaitGroup
+			for i, status := range statuses {
+				// make a decision on each trace
+				if status.State != centralstore.AwaitingDecision {
+					// someone else got to it first
+					continue
+				}
+				stateMap[status.TraceID] = status
+
+				wg.Add(1)
+				go func(status *centralstore.CentralTraceStatus, idx int) {
+					defer wg.Done()
+
+					trace, err := c.Store.GetTrace(ctx, status.TraceID)
+					if err != nil {
+						c.Logger.Error().Logf("error getting trace %s: %s", status.TraceID, err)
+						return
+					}
+					traces[idx] = trace
+				}(status, i)
+			}
+			wg.Wait()
+
+			for _, trace := range traces {
+				if trace == nil {
+					continue
+				}
+
+				if trace.Root != nil {
+					c.Metrics.Increment("trace_decision_has_root")
+				} else {
+					c.Metrics.Increment("trace_decision_no_root")
+				}
+
+				var sampler sample.Sampler
+				var found bool
+
+				tr := c.SpanCache.Get(trace.TraceID)
```

Comment:
Oh, whoops, this is a mistake -- we need to be making the trace decision based on the centralspan information, NOT on the values the refinery has in cache. The rest of this function needs to be rewritten to use that central span data.